src/colvar_neuralnetworkcompute.cpp

```diff
@@ -283,27 +262,1003 @@ std::vector<std::vector<double>> neuralNetworkCompute::multiply_matrix(const std
+    for(int i=0;i<m_num_head;i++)//iterate according to heads
+    {
+        qk_mul.push_back(multiply_num(matmul(q_mul[i],transpose(k_mul[i])),1/sqrt(m_d_model/m_num_head)));//shape=(num_head,seq_len_q,seq_len_k)
+    }
+    qk_mul=softmax(qk_mul);
+    std::vector<std::vector<std::vector<double>>> scaled_qkv;
+    for(int i=0;i<m_num_head;i++)
+    {
+        //shape=(num_head, seq_len_v, depth=d_model/num_head)
+        scaled_qkv.push_back(matmul(qk_mul[i],v_mul[i]));
+    }
+    std::vector<std::vector<double>> concate_qkv=reshape(scaled_qkv);//shape=(seq_len, d_model)
+    concate_qkv=matmul(hstack(concate_qkv,buf),vstack(m_o_weights,reshape(m_o_biases,1,m_o_biases.size())));
+    std::vector<double> output=concat(concate_qkv,m_input_size);
+    return output;
+}
+
+std::vector<std::vector<double>> AttentionLayer::computeGradient(const std::vector<double>& input) const {
+    /*
+    * gradient of input[row*seq_len+column]={softmax((X@WQ)_head@(X@WK)_head^T/sqrt(depth))}@{partial(X@WV)_head/partial(x_ij)}+
+    * {partial(softmax((X@WQ)_head@(X@WK)_head^T/sqrt(depth)))/partial(x_ij)}@{(X@WV)_head}
+    * We notify the four part to be calculated as part1, part2, part3, part4. 
+    * part2=partial(X)/partial(x)@WV
+    * notice that split_head(X@W)==X@split_head(W), that's why we calcualate the total derivative before split head. 
+    * 
+    */ 
+    std::vector<std::vector<double>> q=transpose(reshape(input,int(m_d_model),int(m_input_size/m_d_model)));//seq_len*d_model
+    std::vector<std::vector<double>> k=q;//shape=(seq_len_k,d_model)
+    std::vector<std::vector<double>> v=q;//total assignment
+    std::vector<std::vector<double>> buf_1=creatematrix(q.size(),1,1);//all one matrix
+    std::vector<std::vector<double>> q_expand=hstack(q,buf_1);
+    std::vector<std::vector<double>> k_expand=q_expand;
+    std::vector<std::vector<double>> v_expand=q_expand;
+    std::vector<std::vector<std::vector<double>>> q_mul,k_mul,v_mul;
+    std::vector<std::vector<std::vector<double>>> qk_mul(m_num_head,std::vector<std::vector<double>>(0));
+    std::vector<std::vector<std::vector<double>>> qk_mul_sof;
+    std::vector<std::vector<std::vector<double>>> scaled_qkv(m_num_head,std::vector<std::vector<double>>(0));
+    std::vector<std::vector<double>> concate_qkv;
+    std::vector<double> output;
+    std::vector<std::vector<double>> output_grad(m_output_size, std::vector<double>(m_input_size, 0));
+    std::vector<std::vector<double>> X;
+    int seq_len=m_input_size/m_d_model;
+    int depth=m_d_model/m_num_head;
+
+    q=matmul(q_expand,vstack(m_q_weights,reshape(m_q_biases,1,m_q_biases.size())));
+    q_mul=reshape(q,m_num_head,true);//shape=(num_head,seq_len,d_model/num_head)
+
+    k=matmul(k_expand,vstack(m_k_weights,reshape(m_k_biases,1,m_k_biases.size())));
+    k_mul=reshape(k,m_num_head,true);//split head
+
+    v=matmul(v_expand,vstack(m_v_weights,reshape(m_v_biases,1,m_v_biases.size())));
+    v_mul=reshape(v,m_num_head,true);
```

Comment:
I see `k=matmul(q_expand,vstack(m_k_weights,reshape(m_k_biases,1,m_k_biases.size())));` is in `compute`, but here it is `k=matmul(k_expand,vstack(m_k_weights,reshape(m_k_biases,1,m_k_biases.size())));`. Is it `q_expand` or `k_expand`?

Reply:
The vector `q_expand`  have exactly the same value as `k_expand` and `v_expand`, that's the reason why it is called "self" attention layer. Although to be honest, this scheme still has much room for improvement.